File: mlh_parser/tests/test_synthetic_dates.py
"""
Date parsing tests using synthetic .eml fixtures.

These cover edge cases found in production that are not present in the
real-email test corpus:
  - Out-of-range timezone offsets (e.g. +9900, +2400)
  - Missing Date header entirely

For each case the parser should fall back to the Received header date.
"""

from freezegun import freeze_time
import io
import pytest
from dateutil import parser as date_parser

from mlh_parser.parser import parse_and_process_email

from .helpers import list_files_with_extension, map_to_file_extensions


directory = "./synthetic/"
email_files = list_files_with_extension(directory, ".eml")

date_test_files = [
    map_to_file_extensions(email_f, [".date.pytest"]) for email_f in email_files
]


@freeze_time("2025-12-21")
@pytest.mark.parametrize("email_file, date_file", date_test_files)
def test_synthetic_date_fallback(email_file, date_file) -> None:
    mail_bytes = io.open(email_file, mode="rb").read()
    expected_date = date_parser.parse(
        io.open(date_file, mode="r", encoding="utf-8").read().split("\n")[0].strip()
    )

    result = parse_and_process_email(mail_bytes)

    assert result["date"] == expected_date, (
        f"Expected {expected_date}, got {result['date']}"
    )

File: mlh_parser2/tests/test_synthetic_dates.rs
mod common;

use chrono::DateTime;
use common::{list_files_with_extension, map_to_file_extensions, parse_date_file};
use mlh_parser2::parser::parse_and_process_email;
use std::fs;

#[test]
fn test_synthetic_date_fallback() {
    let directory = "./synthetic/";
    let email_files = list_files_with_extension(directory, ".eml");

    let now = DateTime::from_timestamp(1734748800, 0)
        .unwrap()
        .into();

    for email_file in &email_files {
        let fixtures = map_to_file_extensions(email_file, &[".date.pytest"]);
        if fixtures.is_empty() {
            continue;
        }
        let date_file = &fixtures[0];

        if !date_file.exists() {
            continue;
        }

        let mail_bytes = fs::read(email_file).unwrap();
        let expected_date_str = parse_date_file(date_file);
        if expected_date_str.is_empty() {
            continue;
        }
        let expected_date =
            mlh_parser2::date_parser::parse_date_tentative_raw(&expected_date_str);

        let result = parse_and_process_email(&mail_bytes, now).unwrap();

        if let Some(actual_str) = result.headers.get("date")
            && let Ok(actual) = DateTime::parse_from_rfc3339(actual_str)
                && let Some(expected) = expected_date {
                    assert_eq!(
                        actual, expected,
                        "Date mismatch for {:?}: expected {}, got {}",
                        email_file, expected_date_str, actual_str
                    );
                }
    }
}
